runtime/src/accounts_db.rs

```diff
@@ -1984,6 +2202,211 @@ impl AccountsDB {
+            slot_cache.mark_slot_frozen();
+            slot_cache.report_slot_store_metrics();
+        }
+        self.accounts_cache.report_size();
+    }
+
+    // Force flush the cached roots, flush any unrooted frozen slots as well if there are
+    // > MAX_CACHE_SLOTS of them.
+    pub fn force_flush_accounts_cache(&self) {
+        self.flush_accounts_cache(true);
+    }
+
+    pub fn flush_accounts_cache_if_needed(&self) {
+        self.flush_accounts_cache(false);
+    }
+
+    fn flush_accounts_cache(&self, force_flush: bool) {
+        if !force_flush && self.accounts_cache.num_slots() <= MAX_CACHE_SLOTS {
+            return;
+        }
+
+        // Flush all roots
+        let mut flush_roots_elapsed = Measure::start("flush_roots_elapsed");
+        let cached_roots = self.accounts_cache.clear_roots();
+        let mut max_root = None;
+        for root in cached_roots {
+            self.flush_slot_cache(root);
+            if let Some(max_root_slot) = max_root {
+                if root > max_root_slot {
+                    max_root = Some(root);
+                }
+            } else {
+                max_root = Some(root);
+            }
+        }
+        flush_roots_elapsed.stop();
+
+        // Remove all unrooted slots less than `max_root`, as those are no
+        // longer relevant. See comment in BankForks::set_root() to see why this is safe
+        // to do.
+        if let Some(max_root) = max_root {
+            let removed_slots = self.accounts_cache.remove_slots_le(max_root);
+            for (removed_slot, slot_cache) in removed_slots {
+                // All roots less than `max_root` should have been flushed, so
+                // we shouldn't see any here.
+                assert!(!self.accounts_index.is_root(removed_slot));
+                // Purge the state like accounts index, hashes for unrooted slots
+                let reclaims = self.purge_slot_cache_keys(removed_slot, slot_cache);
+                for (reclaimed_slot, reclaimed_account_info) in reclaims {
+                    assert_eq!(reclaimed_slot, removed_slot);
+                    assert_eq!(reclaimed_account_info.store_id, Self::cache_store_id());
+                }
+            }
+        }
+
+        // If there are > MAX_CACHE_SLOTS, then flush the excess ones to storage
+        for old_slot in self.accounts_cache.find_older_frozen_slots(MAX_CACHE_SLOTS) {
+            self.flush_slot_cache(old_slot);
+        }
+
+        inc_new_counter_info!("flush_roots_elapsed", flush_roots_elapsed.as_us() as usize);
+    }
+
+    fn flush_slot_cache(&self, slot: Slot) {
+        info!("flush_slot_cache slot: {}", slot);
+        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+            let iter_items: Vec<_> = slot_cache.iter().collect();
+            let mut total_size = 0;
+            let (accounts, hashes): (Vec<(&Pubkey, &Account)>, Vec<Hash>) = iter_items
```

Comment:
this is my gut sense; but I think selectively committing alive accounts while consulting to AccountsIndex could save IO considerably. From the actual transaction pattern, it seems that most of accounts are overwritten by newer slots.

I know the current pr is indirectly relying on the shrinker code. But, I think wiring from these to the shrinker saves extra IO further.

Reply:
```
[2020-12-21T22:26:01.742635364Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641064): skip_shrink: false count: 8/1665 byte: 6727830/230686720
[2020-12-21T22:26:01.876901736Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641020): skip_shrink: false count: 17/1975 byte: 24679426/192937984
[2020-12-21T22:26:02.020586266Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640951): skip_shrink: false count: 9/610 byte: 24054754/71303168
[2020-12-21T22:26:02.133616923Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640949): skip_shrink: false count: 7/383 byte: 32885822/79691776
[2020-12-21T22:26:02.307890824Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641032): skip_shrink: false count: 12/1829 byte: 33990687/188743680
[2020-12-21T22:26:02.448433253Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641017): skip_shrink: false count: 10/1069 byte: 12389682/176160768
[2020-12-21T22:26:02.579498451Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641011): skip_shrink: false count: 19/2638 byte: 35577637/314572800
[2020-12-21T22:26:02.745299100Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641019): skip_shrink: false count: 9/1073 byte: 12489748/96468992
[2020-12-21T22:26:02.912640409Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640953): skip_shrink: false count: 8/513 byte: 3294186/92274688
[2020-12-21T22:26:03.029673092Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640955): skip_shrink: false count: 8/863 byte: 31744292/83886080
[2020-12-21T22:26:03.151679845Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641061): skip_shrink: false count: 6/545 byte: 8690259/113246208
[2020-12-21T22:26:03.266149933Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640994): skip_shrink: false count: 7/430 byte: 13068254/79691776
[2020-12-21T22:26:03.377959751Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641023): skip_shrink: false count: 7/1114 byte: 14195199/146800640
[2020-12-21T22:26:03.498594397Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641025): skip_shrink: false count: 7/570 byte: 31474964/71303168
[2020-12-21T22:26:03.627883128Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641001): skip_shrink: false count: 47/2012 byte: 27988030/234881024
[2020-12-21T22:26:03.865587975Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641006): skip_shrink: false count: 8/615 byte: 36081511/83886080
[2020-12-21T22:26:03.984739839Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641037): skip_shrink: false count: 8/1025 byte: 25024381/159383552
[2020-12-21T22:26:04.113945015Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640948): skip_shrink: false count: 17/1516 byte: 24088137/117440512
[2020-12-21T22:26:04.253889187Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641045): skip_shrink: false count: 11/584 byte: 17653222/104857600
[2020-12-21T22:26:04.369249852Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641026): skip_shrink: false count: 9/1189 byte: 12019951/96468992
[2020-12-21T22:26:04.486046652Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641062): skip_shrink: false count: 11/502 byte: 11850219/113246208
[2020-12-21T22:26:04.600162719Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56641016): skip_shrink: false count: 11/2047 byte: 22630421/268435456
[2020-12-21T22:26:04.737644265Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640993): skip_shrink: false count: 23/1993 byte: 38209972/192937984
[2020-12-21T22:26:04.874701796Z INFO  solana_runtime::accounts_db] shrink_stale_slot (56640972): skip_shrink: false count: 12/1668 byte: 20312500/125829120
```

Reply:
the really alive vs shrinked total count comparison:

```
ryoqun@ubuqun:~/work/solana/solana$ xsel | awk '{print $9}' | sed 's|/.*||' | paste -s -d + | bc
291
==> 2020-12-24 16:44:38.908016991|0
ryoqun@ubuqun:~/work/solana/solana$ xsel | awk '{print $9}' | sed 's|.*/||' | paste -s -d + | bc
28428
````

So, about write 100x amplification..